kf/bpf.go

```diff
@@ -274,42 +287,42 @@ func (b *BPF) Stop(ifaceName, direction string, chain bool) error {
 			}
 			b.Cmd = nil
 		}
+	} else if len(b.Program.CmdStop) > 0 && b.Program.UserProgramDaemon {
+		cmd := filepath.Join(b.FilePath, b.Program.CmdStop)
 
-		if err := b.VerifyCleanupMaps(chain); err != nil {
-			log.Error().Err(err).Msgf("stop user program - failed to remove map files %s", b.Program.Name)
-			return fmt.Errorf("stop user program - failed to remove map files %s", b.Program.Name)
+		if err := assertExecutable(cmd); err != nil {
+			return fmt.Errorf("no executable permissions on %s - error %v", b.Program.CmdStop, err)
 		}
 
-		return nil
-	}
-
-	cmd := filepath.Join(b.FilePath, b.Program.CmdStop)
+		args := make([]string, 0, len(b.Program.StopArgs)<<1)
+		args = append(args, "--iface="+ifaceName)     // detaching from iface
+		args = append(args, "--direction="+direction) // xdpingress or ingress or egress
 
-	if err := assertExecutable(cmd); err != nil {
-		return fmt.Errorf("no executable permissions on %s - error %v", b.Program.CmdStop, err)
-	}
-
-	args := make([]string, 0, len(b.Program.StopArgs)<<1)
-	args = append(args, "--iface="+ifaceName)     // detaching from iface
-	args = append(args, "--direction="+direction) // xdpingress or ingress or egress
+		for k, val := range b.Program.StopArgs {
+			if v, ok := val.(string); !ok {
+				err := fmt.Errorf("stop args is not a string for the bpf program %s", b.Program.Name)
+				log.Error().Err(err).Msgf("failed to convert stop args value into string for program %s", b.Program.Name)
+				return err
+			} else {
+				args = append(args, "--"+k+" ="+v)
+			}
+		}
 
-	for k, val := range b.Program.StopArgs {
-		if v, ok := val.(string); !ok {
-			err := fmt.Errorf("stop args is not a string for the ebpf program %s", b.Program.Name)
-			log.Error().Err(err).Msgf("failed to convert stop args value into string for program %s", b.Program.Name)
-			return err
-		} else {
-			args = append(args, "--"+k+"="+v)
+		log.Info().Msgf("bpf program stop command : %s %v", cmd, args)
+		prog := execCommand(cmd, args...)
+		if err := prog.Run(); err != nil {
+			log.Warn().Err(err).Msgf("l3afd : Failed to stop the program %s", b.Program.CmdStop)
 		}
+		b.Cmd = nil
 	}
 
-	log.Info().Msgf("bpf program stop command : %s %v", cmd, args)
-	prog := execCommand(cmd, args...)
-	if err := prog.Run(); err != nil {
-		log.Warn().Err(err).Msgf("l3afd/nf : Failed to stop the program %s", b.Program.CmdStop)
+	// unload the BPF programs
+	if b.ProgMapCollection != nil {
+		if err := b.UnloadProgram(ifaceName, direction); err != nil {
+			return fmt.Errorf("BPFProgram %s unload failed on interface %s with error: %v", b.Program.Name, ifaceName, err)
+		}
+		log.Info().Msgf("%s => %s direction => %s - program is unloaded/detached successfully", ifaceName, b.Program.Name, direction)
 	}
-	b.Cmd = nil
-
 	if err := b.VerifyCleanupMaps(chain); err != nil {
 		log.Error().Err(err).Msgf("stop user program - failed to remove map files %s", b.Program.Name)
```

Comment:
Is this error message correct if we only unload programs directly?

if b.ProgMapCollection != nil